File: src/main/java/ms/konovalov/intellij/hidpi/FontProfileManager.java
package ms.konovalov.intellij.hidpi;

import com.intellij.ide.ui.LafManager;
import com.intellij.ide.ui.UISettings;
import com.intellij.openapi.actionSystem.impl.ActionToolbarImpl;
import com.intellij.openapi.editor.EditorFactory;
import com.intellij.openapi.editor.colors.EditorColorsManager;
import com.intellij.openapi.editor.colors.EditorColorsScheme;

import java.util.Objects;

public class FontProfileManager {

    static void applyProfile(FontProfile profile) {
        UISettings uiSettings = UISettings.getInstance();
        uiSettings.setFontSize(profile.getGlobalFontSize());
        if (profile.getGlobalFontFamily() != null) {
            uiSettings.setFontFace(profile.getGlobalFontFamily());
        }
        uiSettings.setOverrideLafFonts(profile.isOverrideGlobalFont());

        EditorColorsScheme globalScheme = EditorColorsManager.getInstance().getSchemeForCurrentUITheme();
        globalScheme.setEditorFontSize(profile.getEditorFontSize());
        if (profile.getEditorFontFamily() != null) {
            globalScheme.setEditorFontName(profile.getEditorFontFamily());
        }
        globalScheme.setConsoleFontSize(profile.getConsoleFontSize());
        if (profile.getConsoleFontFamily() != null) {
            globalScheme.setConsoleFontName(profile.getConsoleFontFamily());
        }
        applySettings(uiSettings);
    }

    private static void applySettings(UISettings uiSettings) {
        uiSettings.fireUISettingsChanged();
        EditorFactory.getInstance().refreshAllEditors();
        LafManager.getInstance().updateUI();
        ActionToolbarImpl.updateAllToolbarsImmediately();
    }

    @SuppressWarnings("SameParameterValue")
    static FontProfile readCurrentProfile(String name, boolean active) {
        UISettings instance = UISettings.getInstance();
        EditorColorsScheme globalScheme = EditorColorsManager.getInstance().getGlobalScheme();
        return new FontProfile(
                active,
                name,
                instance.getOverrideLafFonts(),
                instance.getFontSize(),
                instance.getFontFace(),
                globalScheme.getEditorFontSize(),
                globalScheme.getEditorFontName(),
                globalScheme.getConsoleFontSize(),
                globalScheme.getConsoleFontName()
        );
    }

    static void deselectAll() {
        FontSizeComponent.getProfiles().forEach(p -> p.setActive(false));
    }

    static boolean changed(FontProfile active, UISettings settings) {
        return !(Objects.equals(settings.getFontSize(), active.getGlobalFontSize())
                && Objects.equals(settings.getFontFace(), active.getGlobalFontFamily())
                && Objects.equals(settings.getOverrideLafFonts(), active.isOverrideGlobalFont()));
    }

    static boolean changed(FontProfile active, EditorColorsScheme settings) {
        return !(Objects.equals(settings.getEditorFontSize(), active.getEditorFontSize())
                && Objects.equals(settings.getEditorFontName(), active.getEditorFontFamily())
                && Objects.equals(settings.getConsoleFontSize(), active.getConsoleFontSize())
                && Objects.equals(settings.getConsoleFontName(), active.getEditorFontFamily()));
    }
}

File: src/main/java/ms/konovalov/intellij/hidpi/FontProfileManager.kt
package ms.konovalov.intellij.hidpi

import com.intellij.ide.ui.LafManager
import com.intellij.ide.ui.UISettings
import com.intellij.openapi.actionSystem.impl.ActionToolbarImpl
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.editor.EditorFactory
import com.intellij.openapi.editor.colors.EditorColorsManager
import com.intellij.openapi.editor.colors.EditorColorsScheme

object FontProfileManager {

    fun applyProfile(profile: FontProfile) {
        val uiSettings = UISettings.instance
        uiSettings.fontSize = profile.globalFontSize
        if (profile.globalFontFamily != null) {
            uiSettings.fontFace = profile.globalFontFamily
        }
        uiSettings.overrideLafFonts = profile.isOverrideGlobalFont

        val globalScheme = EditorColorsManager.getInstance().schemeForCurrentUITheme
        globalScheme.editorFontSize = profile.editorFontSize
        if (profile.editorFontFamily != null) {
            globalScheme.editorFontName = profile.editorFontFamily
        }
        globalScheme.consoleFontSize = profile.consoleFontSize
        if (profile.consoleFontFamily != null) {
            globalScheme.consoleFontName = profile.consoleFontFamily
        }
        applySettings(uiSettings)
    }

    private fun applySettings(uiSettings: UISettings) {
        uiSettings.fireUISettingsChanged()
        EditorFactory.getInstance().refreshAllEditors()
        LafManager.getInstance().updateUI()
        ActionToolbarImpl.updateAllToolbarsImmediately()
    }

    fun readCurrentProfile(name: String, active: Boolean): FontProfile {
        val instance = UISettings.instance
        val globalScheme = EditorColorsManager.getInstance().globalScheme
        return FontProfile(
                active,
                name,
                instance.overrideLafFonts,
                instance.fontSize,
                instance.fontFace,
                globalScheme.editorFontSize,
                globalScheme.editorFontName,
                globalScheme.consoleFontSize,
                globalScheme.consoleFontName
        )
    }

    fun deselectAll() {
        profiles.forEach { p -> p.isActive = false }
    }

    fun changed(active: FontProfile, settings: UISettings): Boolean {
        return !(settings.fontSize == active.globalFontSize
                && settings.fontFace == active.globalFontFamily
                && settings.overrideLafFonts == active.isOverrideGlobalFont)
    }

    fun changed(active: FontProfile, settings: EditorColorsScheme): Boolean {
        return !(settings.editorFontSize == active.editorFontSize
                && settings.editorFontName == active.editorFontFamily
                && settings.consoleFontSize == active.consoleFontSize
                && settings.consoleFontName == active.editorFontFamily)
    }

    val profiles = ApplicationManager.getApplication().getComponent(FontSizeComponent::class.java).state.profiles
}
